### PR TITLE
Bitcoin Currency Exchange Optimization

### PR DESCRIPTION
**Pull request by markm1992**
Gets rates from 4 API's for all available currencies supported in the API's. If a request falls within 15minutes of a Time Stamp then it will pull the local value, after 15minutes it calls the API's again.

### CODE REVIEW

File: js/utils/getBitcoinPrice.js [hunk 1]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
             }
+            btPrices.push(bitCoinAvgCurrencies);
           })
           .fail(function (jqXHR, textStatus, errorThrown)
           {
-            //console.log("blockChain request failed: ");
-            //console.log(jqXHR);
-            //console.log(textStatus);
-            //console.log(errorThrown);
+            console.log("Bit coin average request failed: ");
```
Comment (hoffmabc) [inline js/utils/getBitcoinPrice.js]: Should be Bitcoin not Bit coin

File: js/utils/getBitcoinPrice.js [hunk 2]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
-        //console.log(errorThrown);
+        console.log("blockChain request failed: ");
+        console.log(jqXHR);
+        console.log(textStatus);
+        console.log(errorThrown);
       })
       .always(function ()
       {
-        callCoindesk();
+        callCoinKite();
```
Comment (hoffmabc) [inline js/utils/getBitcoinPrice.js]: Coinkite not CoinKite or coinKite

File: js/utils/getBitcoinPrice.js [hunk 3]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
         }
+        btPrices.push(blockChainCurrencies);
       })
       .fail(function (jqXHR, textStatus, errorThrown)
       {
-        //console.log("bitcoinAverage request failed:");
-        //console.log(jqXHR);
-        //console.log(textStatus);
-        //console.log(errorThrown);
+        console.log("blockChain request failed: ");
```
Comment (hoffmabc) [inline js/utils/getBitcoinPrice.js]: Blockchain or blockchain but not blockChain.

File: js/utils/getBitcoinPrice.js [hunk 4]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
-            //console.log(errorThrown);
+            console.log("coinKite request failed: ");
+            console.log(jqXHR);
+            console.log(textStatus);
+            console.log(errorThrown);
           })
           .always(function ()
           {
-            callBlockchain();
+            callBitCoinAvg();
```
Comment (hoffmabc) [inline js/utils/getBitcoinPrice.js]: callBitcoinAvg not callBitCoinAvg.

**Comment by hoffmabc** [comment]
Sorry for being a stickler here but we just need to be consistent with the case of companies and Bitcoin itself. @jjeffryes can review for functionality. Thanks a lot for the submission! Good job.

File: js/utils/getBitcoinPrice.js [hunk 5]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
+        method: "GET",
+        url: "https://blockchain.info/ticker"
+      })
       .done(function (response)
       {
-        //console.log("bitcoinAverage: " + response['24h_avg']);
-        if($.isNumeric(response['24h_avg'])) {
-          btPrices.push(response['24h_avg']);
+        var blockChainCurrencies = {};
+        for (var currency in response) {
```
Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:30]: On all of your for in loops, wrap the inside in an if statement that makes sure the properties belong to the object, not the object's prototype. It probably works fine in this case, but this is a JSHint check we require before code is merged. 

For example:

for (var blockChainCurrency in response) {
  if (response.hasOwnProperty(blockChainCurrency)) {
    ...do things
  }
}

https://jslinterrors.com/the-body-of-a-for-in-should-be-wrapped-in-an-if-statement

Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:30]: Don't re-use currency as a variable name, it makes it confusing as to whether you're referring to the loop's currency var or the currency parameter that is part of the parent scope. Use something specific to this loop, like blockChainCurrency.

File: js/utils/getBitcoinPrice.js [hunk 6]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
-        url: "https://api.coindesk.com/v1/bpi/currentprice/" + currency + ".json"
+        url: "https://api.coinkite.com/public/rates"
       })
           .done(function (response)
           {
-            //console.log("coinDesk: " + response.bpi[currency]['rate']);
-            if($.isNumeric(response.bpi[currency].rate)) {
-              btPrices.push(response.bpi[currency].rate);
+            var coinKiteCurrencies = {};
+            for (var currency in response.rates.BTC) {
```
Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:57]: See comments above.

File: js/utils/getBitcoinPrice.js [hunk 7]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
+        url: "https://api.bitcoinaverage.com/all"
       })
           .done(function (response)
           {
-            //console.log("blockChain: " + response[currency]['15m']);
-            if($.isNumeric(response[currency]['15m'])) {
-              btPrices.push(response[currency]['15m']);
+            var bitCoinAvgCurrencies = {};
+            for (var currency in response) {
+                if(response[currency].averages) {
```
Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:86]: See comment above on currency. In this case your if statement satisfies the need to make sure the property you're looping over belongs to the response object, so you don't need to change that.

File: js/utils/getBitcoinPrice.js (hunk 5 above)
Comment (markm1992) [inline js/utils/getBitcoinPrice.js:30]: Good note, my jsHint didn't pick up on that error. Will fix it and upload soon.

File: js/utils/getBitcoinPrice.js [hunk 8]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
-        url: "https://api.coinkite.com/public/rates"
+        url: "http://api.bitcoincharts.com/v1/weighted_prices.json"
       })
           .done(function (response)
           {
-            //console.log("coinKite: " + response.rates.BTC[currency]['rate']);
-            if ($.isNumeric(response.rates.BTC[currency].rate)) {
-              btPrices.push(response.rates.BTC[currency].rate);
+            var bitCoinChartsCurrencies = {};
+            for (var currency in response) {
```
Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:114]: See comments on loops above.

File: js/utils/getBitcoinPrice.js [hunk 9]
```diff
@@ -13,107 +13,115 @@ module.exports = function(currency, callback){
...
-    var callCoinKite = function ()
+    var callBitcoinCharts = function ()
     {
       $.ajax({
         method: "GET",
-        url: "https://api.coinkite.com/public/rates"
+        url: "http://api.bitcoincharts.com/v1/weighted_prices.json"
       })
           .done(function (response)
           {
```
Comment (jjeffryes) [inline js/utils/getBitcoinPrice.js:112]: bitcoincharts doesn't return JSON, it returns a string, which was causing all the currencies to come back as undefined below. 

For all of the AJAX calls, it's a good idea to check if the response is in JSON, and if not do JSON.parse() on the response before trying to pull data out of it.

**Comment by jjeffryes** [comment]
Fix the issues mentioned above, and then resubmit the pull request.

**Pull request closed without merging**